< [GMIT Data Analytics](https://web.archive.org/web/20201029063153/https://www.gmit.ie/computer-science-and-applied-physics/higher-diploma-science-computing-data-analytics-ict) | [Table of Contents](https://github.com/E6985) | [README](https://github.com/E6985/fda-tasks/blob/master/README.md) >

<img src="img/gmit.png" />

## Fundamentals of Data Analysis - Tasks 2020
## Due: last commit on or before December 18th, 2020
***

#### Task 1
Write a Python function called counts that takes a list as input and returns a dictionary of unique items in the list as keys and the number of times each item appears as values. So, the input ``['A', 'A', 'B', 'C', 'A']`` should have output ``{'A': 3, 'B': 1, 'C': 1}``. Your code should not depend on any module from the standard library or otherwise. By the standard library, we mean the modules and packages that come as standard with Python. Anything built-in that can be used without an import statement can be used. You should research the task first and include a description with references of your algorithm in the notebook.
***

###### Dictionary
***

- dictionary - unorderd data collection type - flexible mappings of keys to values - form the basis of much of pythons internal implementation - created via a comma-separated list - pairs - ``key:value`` - within curly braces [
[...] J. VanderPlas, "Built-In Data Structures," A Whirlwind Tour of Python, p. 60, August 2016.
]

###### Algorithm - Attempt - Initial Approach
***

- input - list - ``['A', 'A', 'B', 'C', 'A']`` - function - counts - should take any list containing any items and return a dictionary of unique items in the list as keys and the number of times each item appears as values - algorithm designed should be based on imagining the list contains - letters in strings as presented in the Tasks 1 - list containing something other than strings - containing a single character - containing numbers - containing other lists - containing dictionaries - **think about this** - ok to **specify if the the function does not work with** lists of lists or something like that but consider said ideas - consider all of the different ins and outs where the algorithm might fail and document - have a lot of freedom here [
[...] I. McLoughlin, "Tasks 2020 Fundamentals of Data Analysis," GMIT, October 2020.
]

In [1]:
nArgList=['A', 'A', 'B', 'C', 'A']

In [2]:
# define function - counts
def counts(nParList):
    nDictionary={}
    for nEachValue in nParList:
        # object Dict - method setdefault - call
        nDictionary.setdefault(nEachValue,0)
        nDictionary[nEachValue]+=1
    return nDictionary

In [3]:
# function counts - call
counts(nArgList)

{'A': 3, 'B': 1, 'C': 1}

###### Python Module - collections
***

- built-in - module - collections - provides alternatives to pythons general purpose built-in container - dict [
[...] Python Documentation, "Collections - Container dataypes," [Python](https://web.archive.org/web/20201022092017/https://docs.python.org/3/library/collections.html#module-collections), October 2008.
]

In [4]:
# a_S_tandard_L_ibrary-imports
from collections import Counter as Cctr

- module - collections - class - Counter - elements are stored as dictionary keys - item counts are stored as dictionary values [
[...] Python Documentation, "Collections - Container dataypes," [Python](https://web.archive.org/web/20201022092017/https://docs.python.org/3/library/collections.html#module-collections), October 2008.
] - this task can depend on any module from the standard library or otherwise - ``collections`` - should not be used - the idea here is to write code from scratch - putting some decent constraints on the input - reviewing module - collections - to get a feel for how it works [
[...] I. McLoughlin, "Tasks 2020 Fundamentals of Data Analysis," GMIT, October 2020.
]

In [5]:
# module collections - object Counter alias Cctr - instantiate
Cctr(nArgList)

Counter({'A': 3, 'B': 1, 'C': 1})

###### Multiset Theory
***

- implementation based on - multiset theory - shortened to - mset - multiple-membership set - common accepted name replacing - bag - bunch - weighted set - occurrence set - heap - sample - fireset - modification of - set - collection of objects called elements in which the elements can occur many times - order is ignored - number of elements occurring in a multiset is called - multiplicity [
[...] W. Blizard, "Multiset theory," [Project euclid](https://web.archive.org/web/20201105163941/https://projecteuclid.org/euclid.ndjfl/1093634995), July 1987.
]

- multiplicity is explicitly significant - precise number of elements [
[...] E. Weisstein, "Multiset," [mathworld.wolfram](https://web.archive.org/web/20200319012240/https://mathworld.wolfram.com/Multiset.html), 2002.
]

$$\{3,1,2\}\equiv\{1,2,3\}$$

$$\{3,3,1,2\}\ne\{1,2,3\}$$

- the multiplicity of an element is defined as the number of times it occurs in the multiset - denoted by listing its elements separated by commas between curly braces - similiar to pythons unorderd data collection type - dictionary [
[...] J. Nieminen, J. Khim, "Multiset," [Brilliant](https://web.archive.org/web/20201105164908/https://brilliant.org/wiki/multiset), April 2018.
]

$$\{a,a,b,c,a\}$$

- elements are listed with their multiplicities as colon-separated pairs [
[...] J. Nieminen, J. Khim, "Multiset," [Brilliant](https://web.archive.org/web/20201105164908/https://brilliant.org/wiki/multiset), April 2018.
]

$$\{a:3,b:1,c:1\}$$

###### Algorithm - Research - Module collections Approach

- source code - module - collections - Class - Counter - method - most_common [
[...] Python, "cpython/Lib/collections/__init__.py," [Github](https://web.archive.org/web/20201105151531/https://github.com/python/cpython/blob/master/Lib/collections/__init__.py), May 2020.
] - algorithm used - programming language - Smalltalk - Bag.sortedByCount [
[...] Smalltalk, "smalltalk/kernel/Bag.st," [Github](https://web.archive.org/web/20201105154907/https://github.com/gnu-smalltalk/smalltalk/blob/master/kernel/Bag.st), January 2010.
] - line - 568 [
[...] Python, "cpython/Lib/collections/__init__.py," [Github](https://web.archive.org/web/20201105151531/https://github.com/python/cpython/blob/master/Lib/collections/__init__.py), May 2020.
]

In [6]:
%%script bash
cat -n C:/Users/E6985/anaconda3/lib/collections/__init__.py | grep "def most_common" -A 11

   559	    def most_common(self, n=None):
   560	        '''List the n most common elements and their counts from the most
   561	        common to the least.  If n is None, then list all element counts.
   562	
   563	        >>> Counter('abracadabra').most_common(3)
   564	        [('a', 5), ('b', 2), ('r', 2)]
   565	
   566	        '''
   567	        # Emulate Bag.sortedByCount from Smalltalk
   568	        if n is None:
   569	            return sorted(self.items(), key=_itemgetter(1), reverse=True)
   570	        return _heapq.nlargest(n, self.items(), key=_itemgetter(1))


###### Algorithm - Research - Object-Oriented Approach
***

- keyword - class - indicates creation - in python all classes create objects - implicit default - ``class (object)`` - python 3 - ``(object)`` - not required [
[...] D. Amos, "Object-Oriented Programming (OOP) in Python 3," [Real Python](https://web.archive.org/web/20201107154103/https://realpython.com/python3-object-oriented-programming), July 2020.
] - source code - Appendix A - object-oriented implementation of mulitset - bag - whereby - class - Counts - derives from object - dict [
[...] E. Wiethoff, "Bag Sum In Many Programming Languages," [wiki.c2](https://web.archive.org/web/20201107143701/https://wiki.c2.com/?BagSumInManyProgrammingLanguages), n.d.
]

```python
# define class - Counts
class Counts(dict):
```

- method - \_\_init\_\_ - initialise - specify - an objects initial attributes by giving them their default values - state - must have at least one argument - variable - self - refers to the object itself - Counts - the method - \_\_init\_\_ - called automatically when creating a new - Counts - instance [
[...] D. Amos, "Object-Oriented Programming (OOP) in Python 3," [Real Python](https://web.archive.org/web/20201107154103/https://realpython.com/python3-object-oriented-programming), July 2020.
]

```python
# instance attributes - initialiser
def __init__(self,nArgInput):
```

- loop through each element in the input list - method - performCount - called on each element [
[...] No Reference.
]

```python
    for nEachElement in nArgInput:
        self.performCount(nEachElement)
```

- method - performCount - the object - self - set of elements from the input list - assigned the count of each occurrence of the list elements - keeping count - add one to each occurrence [
[...] W3Schools, "Python Dictionary get() Method," [w3schools](https://web.archive.org/web/20201107161412/https://www.w3schools.com/python/ref_dictionary_get.asp), August 2018.
] 

```python
# instance attributes - initialiser
def performCount(self,nEachElement):
    self[nEachElement]=self.get(nEachElement,0)+1
```

- method __dict__ - single line - list comprehension - compressing a list-building - for - loop into a single short readable line [
[...] J. VanderPlas, "Built-In Data Structures," A Whirlwind Tour of Python, p. 58, August 2016.
] 


- each ``key`` - ``value`` - ìn the newly returned sorted set from the elements in the iterable - object - self - looping through the dictionary - the key plus corresponding value - retrieved at the same time using the method - items [
[...] Python Documentation, "5.6. Looping Techniques," [Python](https://web.archive.org/web/20201107162523/https://docs.python.org/3/tutorial/datastructures.html#looping-techniques), October 2008.
]
    
```python
def __dict__(self):
    return [(nEachKey,nEachvalue) for (nEachKey,nEachvalue) in sorted(self.items())]
```

- reviewing the implementation from the module - collections - can see similarities - line - 569 [
[...] Python, cpython/Lib/collections/__init__.py," [Github](https://web.archive.org/web/20201105151531/https://github.com/python/cpython/blob/master/Lib/collections/__init__.py), May 2020.
]

In [7]:
%%script bash
cat -n C:/Users/E6985/anaconda3/lib/collections/__init__.py | grep "# Emulate Bag.sortedByCount from Smalltalk" -A 2

   567	        # Emulate Bag.sortedByCount from Smalltalk
   568	        if n is None:
   569	            return sorted(self.items(), key=_itemgetter(1), reverse=True)


In [8]:
# define class - Counts
class Counts(dict):
    # instance attributes - initialiser
    def __init__(self,nArgInput):
        for nEachElement in nArgInput:
            self.performCount(nEachElement)
    # instance attributes - initialiser
    def performCount(self,nEachElement):
        self[nEachElement]=self.get(nEachElement,0)+1
    # instance attributes - initialiser
    def __dict__(self):
        return [(nEachKey,nEachvalue) for (nEachKey,nEachvalue) in sorted(self.items())]

In [9]:
# object Counts - instantiate
Counts(nArgList)

{'A': 3, 'B': 1, 'C': 1}

###### Algorithm - Research - Eliminate Duplicate Elements Approach
***

- source code - Appendix B - list comprehension -  implementation of mulitset - bag - uses a dictionary to get rid of duplicate elements in the list to be counted [
[...] E. Wiethoff, "Bag Sum In Many Programming Languages," [wiki.c2](https://web.archive.org/web/20201107143701/https://wiki.c2.com/?BagSumInManyProgrammingLanguages), n.d.
]

- list comprehension - for - loop - readable line [
[...] J. VanderPlas, "Built-In Data Structures," A Whirlwind Tour of Python, p. 58, August 2016.
] - module - array - method - count - returns the number of occurrences of each element in the array [
[...] Python Documentation, "array — Efficient arrays of numeric values," [Python](https://web.archive.org/web/20201108173847/https://docs.python.org/3/library/array.html?highlight=count#array.array.count), October 2008.
]

In [10]:
[(nEachElement,nArgList.count(nEachElement)) for nEachElement in nArgList]

[('A', 3), ('A', 3), ('B', 1), ('C', 1), ('A', 3)]

- constructor - dict - create a new dictionary - if a key occurs more than once then the last value for that key becomes the corresponding value in the new dictionary - eliminating duplicate elements [
[...] Python Documentation, "Built-in Types," [Python](http://web.archive.org/web/20201108174452/https://docs.python.org/3/library/stdtypes.html#dict), October 2008.
]

In [11]:
dict([(nEachElement,nArgList.count(nEachElement)) for nEachElement in nArgList])

{'A': 3, 'B': 1, 'C': 1}

In [12]:
# define function - counts
def counts(nParList):
    return dict([(nEachElement,nArgList.count(nEachElement)) for nEachElement in nArgList])

In [13]:
# function counts - call
counts(nArgList)

{'A': 3, 'B': 1, 'C': 1}

###### Algorithm - Research - Sort and Tally Approach
***

- source code - Appendix C - no-frills implementation - little use of functions/methods - only method used - sorted - simple sort and tally algorithm [
[...] Anonymous, "Bag Sum In Many Programming Languages," [wiki.c2](https://web.archive.org/web/20201107143701/https://wiki.c2.com/?BagSumInManyProgrammingLanguages), n.d.
]

- implementation from the module - collections - line - 569 [
[...] Python, cpython/Lib/collections/__init__.py," [Github](https://web.archive.org/web/20201105151531/https://github.com/python/cpython/blob/master/Lib/collections/__init__.py), May 2020.
] - uses the function - sorted - as part of its approach - this returns a new sorted list from the items in the iterable [
[...] Python Documentation, "Built-in Functions," [Python](https://web.archive.org/web/20200512035500/https://docs.python.org/3/library/functions.html#sorted), October 2008.
] 

In [14]:
%%script bash
cat -n C:/Users/E6985/anaconda3/lib/collections/__init__.py | grep "# Emulate Bag.sortedByCount from Smalltalk" -A 2

   567	        # Emulate Bag.sortedByCount from Smalltalk
   568	        if n is None:
   569	            return sorted(self.items(), key=_itemgetter(1), reverse=True)


In [15]:
nArgList

['A', 'A', 'B', 'C', 'A']

- method - sort - modifies the list in-place [
[...] Python Documentation, "Built-in Types," [Python](https://web.archive.org/web/20200502135931/https://docs.python.org/3/library/stdtypes.html#list.sort), October 2008.
]

In [16]:
nArgList.sort()
nArgList

['A', 'A', 'A', 'B', 'C']

- initialise result - name - nDictionary - empty dictionary [
[...] No Reference.
]

In [17]:
nDictionary={}
nDictionary

{}

- initialise - name - nListItem - zeroth element in the input list - nArgList [
[...] No Reference.
]

In [18]:
nListItem=nArgList[0]
nListItem

'A'

- initialise - name - nListItemCount - to one given the zeroth element - first element - in the input list - nArgList - will start at count - ``1`` [
[...] No Reference.
]

In [19]:
nListItemCount=1
nListItemCount

1

- accessing elements in compound types - ``dict`` - done through indexing for single elements - slicing for multiple elements - both indicated by a square-bracket syntax [
[...] J. VanderPlas, "Built-In Data Structures," A Whirlwind Tour of Python, p. 32, August 2016.
] using a for loop iterating through the input list - nArgList - starting at position index - ``1`` - given that we have already determined position index - ``0`` - is the letter - ``A`` - after the input list - nArgList - was sorted via the object - list - method - sort [
[...] No Reference.
]

```python
for nEachValue in nArgList[1:]:
```

- conditional - if - the element under investigation - name - nEachValue - is the same as element assigned to - name - nListItem - then add one to the - name - nListCount - said element now has a count - ``2`` - in addition start populating the result - dictionary - name - nDictionary - via indexing set the first - ``key`` - to its count - ``value`` - name - nListItemCount [
[...] No Reference.
]

```python
if nEachValue==nListItem:
    nListItemcount+=1
    nDictionary[nEachValue]=nListItemcount
```

- conditional - else - initialise - name - nListItem - to the - name - nEachValue - element under investigation from the input list - nArgList - initialise - name - nListItemCount - restarting the count - ``1`` - continue populating the resulting dictionary - name - nDictionary - via indexing set the current element under investigation - name - nEachValue - ``key`` - to its current count - ``value`` - name - nListItemCount [
[...] No Reference.
]

```
else:
    nListItem,nListItemcount=nEachValue,1
    nDictionary[nEachValue]=nListItemcount
```

- return result - nDictionary

```python
return nDictionary
```

In [20]:
nArgList=['A', 'A', 'B', 'C', 'A'] # reset list - sorted earlier 
# define function - counts
def counts(nParList):
    nArgList.sort()
    nDictionary={}
    nListItem,nListItemcount=nArgList[0],1
    for nEachValue in nArgList[1:]:
        if nEachValue==nListItem:
            nListItemcount+=1
            nDictionary[nEachValue]=nListItemcount
        else:
            nListItem,nListItemcount=nEachValue,1
            nDictionary[nEachValue]=nListItemcount
    return nDictionary

In [21]:
# function counts - call
counts(nArgList)

{'A': 3, 'B': 1, 'C': 1}

###### Input - Lists - Test Function counts
***

- list - data structure - standard mutable - changeable - multielement container [
[...] Jake VanderPlas, "A Python List Is More Than Just a List," Python Data Science Handbook, December 2016.
]

- list comprehension - class - type - returns object type - everything in python is an object [
[...] Python Documentation, "Built-in Functions," [Python](https://web.archive.org/web/20201109201230/https://docs.python.org/3/library/functions.html#type), October 2008.
]

- function - fSummary - for display purposes only - not form part of submission [
[...] No Reference.
]

In [22]:
# define function - fSummary
def fSummary(nParList):
    print("<LIST>   {}".format(nParList))
    print("<LENGTH>  {}".format(len(nParList)))
    print("<TYPES>  ")
    return [type(nEachElement) for nEachElement in nParList]

- using most commonly used iterator in python - range - generate a sequence of numbers - ``0`` - ``1`` - by convention the top of the object range is not included in the output - append numbers - ``2`` - ``3`` - ``2`` - list has five elements - four distinct - all of type - ``int`` [
[...] J. VanderPlas, "Built-In Data Structures," A Whirlwind Tour of Python, p. 39, August 2016.
]

In [23]:
nArgList=list(range(2))
nArgList.append(2)
nArgList.append(3)
nArgList.append(2)
# call function - fSummary
fSummary(nArgList)

<LIST>   [0, 1, 2, 3, 2]
<LENGTH>  5
<TYPES>  


[int, int, int, int, int]

In [24]:
# function counts - call
counts(nArgList)

{1: 1, 2: 2, 3: 1}

&#x2623; **key:value - pairs - ``4`` - pair associated with zero - ``0`` - no count completed** &#x2623;

In [25]:
# call function - fSummary
fSummary(counts(nArgList))

<LIST>   {1: 1, 2: 2, 3: 1}
<LENGTH>  3
<TYPES>  


[int, int, int]

- list comprehension - cast all elements in existing list - nArgList - to strings [
[...] No Reference.
]

In [26]:
nArgList=[str(nEachElement) for nEachElement in nArgList]
# call function - fSummary
fSummary(nArgList)

<LIST>   ['0', '1', '2', '2', '3']
<LENGTH>  5
<TYPES>  


[str, str, str, str, str]

&#x2623; **key:value pairs - ``4`` - pair associated with string zero - ``"0"`` - no count completed** &#x2623; 

In [27]:
# call function - fSummary
fSummary(counts(nArgList))

<LIST>   {'1': 1, '2': 2, '3': 1}
<LENGTH>  3
<TYPES>  


[str, str, str]

- list comprehension - cast all elements in existing list - nArgList - to floats [
[...] No Reference.
]

In [28]:
nArgList=[float(nEachElement) for nEachElement in nArgList]
# call function - fSummary
fSummary(nArgList)

<LIST>   [0.0, 1.0, 2.0, 2.0, 3.0]
<LENGTH>  5
<TYPES>  


[float, float, float, float, float]

In [29]:
# function counts - call
counts(nArgList)

{1.0: 1, 2.0: 2, 3.0: 1}

&#x2623; **key:value pairs - ``4`` - pair associated with zero - ``0.0`` - no count completed** &#x2623; 

In [30]:
# call function - fSummary
fSummary(counts(nArgList))

<LIST>   {1.0: 1, 2.0: 2, 3.0: 1}
<LENGTH>  3
<TYPES>  


[float, float, float]

- creating heterogeneous lists - diverse content - data types - floats - appending single integer [
[...] Jake VanderPlas, "A Python List Is More Than Just a List," Python Data Science Handbook, December 2016.
]

In [31]:
nArgList.append(1)
# call function - fSummary
fSummary(nArgList)

<LIST>   [0.0, 1.0, 2.0, 2.0, 3.0, 1]
<LENGTH>  6
<TYPES>  


[float, float, float, float, float, int]

In [32]:
# function counts - call
counts(nArgList)

{1.0: 2, 2.0: 2, 3.0: 1}

&#x2623; **key:value pairs - ``5`` - pair associated with zero - ``0.0`` - no count completed** &#x2623; 

&#x2623; **append - integer ``1`` - output captured under float - ``1.0`` - could argue it works albeit should output pair - key:value - for the single integer - ``1`` - as - ``1:1``** &#x2623; [
[...] No Reference.
]

In [33]:
# call function - fSummary
fSummary(counts(nArgList))

<LIST>   {1.0: 2, 2.0: 2, 3.0: 1}
<LENGTH>  3
<TYPES>  


[float, float, float]

***
#### References
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[1] authorNames, referenceTitle, locationWebsiteBookVideo, dateMonthYear.

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[10]

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[100]

&nbsp;&nbsp;&nbsp;[1000]
***
#### END

#### Task 2
Write a Python function called dicerolls that simulates rolling dice. Your function should take two parameters: the number of dice ``k`` and the number of times to roll the dice ``n``. The function should simulate randomly rolling ``k`` dice ``n`` times, keeping track of each total face value. It should then return a dictionary with the number of times each possible total face value occurred. So, calling the function as ``diceroll(k=2, n=1000)`` should return a dictionary like:

``{2:19,3:50,4:82,5:112,6:135,7:174,8:133,9:114,10:75,11:70,12:36}``

You can use any module from the Python standard library you wish and you should include a description with references of your algorithm in the notebook.
***

- standard six sided dice - each of the six faces labelled - ``1`` - ``2`` - ``3`` - ``4`` - ``5`` - ``6`` - simulate rolling a number of dice a number of times [
[...] I. McLoughlin, "Task 2 Rolling dice," GMIT, November 2020.
]

- might roll two dice - two six sided dice - roll both once - maybe one of the dice face comes up - ``3`` - the other dice face comes up - ``5`` - capturing total - ``8`` - it is this eight - total face value - interested in [
[...] I. McLoughlin, "Task 2 Rolling dice," GMIT, November 2020.
]

- function - dicerolls - two parameters - i) number of dice - ``k`` - ii) number of times to roll the dice - ``n`` - if - ``n`` - 100 - ``k`` - 2 - simulates rolling two dice one-hundred times keeping track of the face values - the function should simulate randomly rolling - ``k`` - dice - ``n`` times - keeping track of each total face value [
[...] I. McLoughlin, "Task 2 Rolling dice," GMIT, November 2020.
]

- return dictionary - number of times each possible total face value occurred [
[...] I. McLoughlin, "Task 2 Rolling dice," GMIT, November 2020.
]

```python
dicerolls(k=2,n=1000)
```

- call - function - dicerolls - rolling - ``2`` - dice - ``1000`` - times should return a dictionary - first key listed - ``2`` - showed up - ``19`` - times - of the one-thousand times the dice was rolled - only one way to get - ``2`` - when the two dice face values are both - ``1`` - ``1`` - know immediately two dices showed up - ``1`` - ``1`` - obviously calling the function should get slightly different results - randomly rolling the dice [
[...] I. McLoughlin, "Task 2 Rolling dice," GMIT, November 2020.
]

``{2:19,3:50,4:82,5:112,6:135,7:174,8:133,9:114,10:75,11:70,12:36}``

- might be able to pick out some trends - analyse those trends to see whats happening - interesting to simulate rolling dice to see patterns - see that the lower numbers and the higher numbers seem to be less likely than the number bunched around the middle - can ask yourself why that might be - suggestion - only one way to get - ``2`` - thats by rolling - ``1`` - ``1`` - might be many ways to get - ``6`` [
[...] I. McLoughlin, "Task 2 Rolling dice," GMIT, November 2020.
]

- roll only one dice - if random function is working correctly - should see each of the six sides with equal probability - should see each of them roughly the same number of times when you start adding the faces together - all of a sudden you lose uniformity [
[...] I. McLoughlin, "Task 2 Rolling dice," GMIT, November 2020.
]

- can use any python module from the python standard library - can use modules that are not in the standard library but for this task maybe not needed - use the random module or something like that in python - probably a short description but might get into having a look at some results - analysing the results to beef up the Jupyter Notebook - maybe even a histogram of the results [
[...] I. McLoughlin, "Task 2 Rolling dice," GMIT, November 2020.
]

In [1]:
k=2

In [2]:
n=10

In [3]:
from random import randint as r
def dicerolls(k,n):
    nList=[]
    # t=0
    for i in range(n):
        # t+=(r(1,6)*k)
        nList.append(r(1,6)*k)
        if len(nList)==n:
            nList.sort()     

    print(nList,end="\n\n")

    return dict([(nEachElement,nList.count(nEachElement)) for nEachElement in nList])

In [4]:
print(dicerolls(k,n))

[2, 2, 2, 4, 6, 6, 6, 8, 8, 10]

{2: 3, 4: 1, 6: 3, 8: 2, 10: 1}


###### Probability
***

- computing - bit - variable that can take on one of two values - ``1`` - ``0`` - unless some information is presented ahead of time about what the bit contains then can assume to get - ``1`` - half the time - ``0`` - half the time [
[...] I. McLoughlin, "Coin flipping in Python," GMIT, September 2019.
]

- coin flip - one of two outcomes - head - ``1`` - tail - ``1`` - fair coin - normal to assume flip outcome equally likely - give head - ``50%`` - of the time - give tail - ``50%`` - of the time [
[...] I. McLoughlin, "Coin flipping in Python," GMIT, September 2019.
]

- dice roll - one of six outcomes - ``1`` - ``2`` - ``3`` - ``4`` - ``5`` - ``6`` - fair dice - normal to assume roll outcome approximately - ``16.7%`` - likely - give each face value - ``1`` - ``2`` - ``3`` - ``4`` - ``5`` - ``6`` - ``~16.7%`` - chance [
[...] I. McLoughlin, "Coin flipping in Python," GMIT, September 2019.
]

- generally in mathematics - statistics - probability - data analytics - do not like percentages for uncertain outcomes - percentage just means out of a 100 - something that happens 50% of the time - means that out of 100 times of doing it - 50 of those are likely to be something - flip a coin 100 times - expect about 50 heads [
[...] I. McLoughlin, "Coin flipping in Python," GMIT, September 2019.
]

- prefer to say that the probability of something is a value between zero and one [
[...] I. McLoughlin, "Coin flipping in Python," GMIT, September 2019.
] - interval - ``[0,1]`` - greater than or equal to 0 and less than or equal to 1 [
[...] Wikipedia, "Interval (mathematics)," [wikipedia](http://web.archive.org/web/20200925042254/https://en.wikipedia.org/wiki/Interval_%28mathematics%29), 
reference] n.d.
]

- each face value - ``16.7%`` - chance - prefer - probability - ``0.167`` - obviously in a given scenario - if wondering what the value of something is going to come out at - all of the possible values - probabilities of all - should sum up to - ``1`` - for instance when flipping a coin - know the probability of getting a head is - ``0.5`` - know the only other possible outcome is a tail - then it too must have a probability of - ``0.5`` - because the two probabilities must add up to - ``1`` [
[...] I. McLoughlin, "Coin flipping in Python," GMIT, September 2019.
]

- rolling three dice - intuitively would expect the three dice rolls should come up different the majority of the time - question - chance of all three rolls being different [
[...] M. Haahr, "Frequently Asked Questions (FAQ)," [random](https://web.archive.org/web/20201115135742/https://www.random.org/faq/#Q2.8), n.d.
]

- first die is trival in that any of the six possible values could show - none will result in duplicates [
[...] M. Haahr, "Frequently Asked Questions (FAQ)," [random](https://web.archive.org/web/20201115135742/https://www.random.org/faq/#Q2.8), n.d.
]

In [1]:
"<probability> {:.0f} <duplicates> {:.0f}%".format(6/6,(1-(6/6)))

'<probability> 1 <duplicates> 0%'

- second die coming up different from the first - one less [
[...] M. Haahr, "Frequently Asked Questions (FAQ)," [random](https://web.archive.org/web/20201115135742/https://www.random.org/faq/#Q2.8), n.d.
]

In [2]:
"<probability> {:.0f} <duplicates> {:.2f}%".format(5/6,(1-(6/6)*(5/6))*100)

'<probability> 1 <duplicates> 16.67%'

- third die coming up different from the first and second - two less [
[...] M. Haahr, "Frequently Asked Questions (FAQ)," [random](https://web.archive.org/web/20201115135742/https://www.random.org/faq/#Q2.8), n.d.
]

In [3]:
"<probability> {:.0f} <duplicates> {:.2f}%".format(5/6,(1-(6/6)*(5/6)*(4/6))*100)

'<probability> 1 <duplicates> 44.44%'

- total probability - ``~0.56`` - of all three turning out different [
[...] M. Haahr, "Frequently Asked Questions (FAQ)," [random](https://web.archive.org/web/20201115135742/https://www.random.org/faq/#Q2.8), n.d.
]

$$\frac6{6}*\frac5{6}*\frac4{6}\approx55.55\%$$

## 6

In [4]:
(6/6)*(5/6)*(4/6)*(3/6)*(2/6)*(1/6)

0.015432098765432098

## 5

In [5]:
(6/6)*(5/6)*(4/6)*(3/6)*(2/6)

0.09259259259259259

## 4

In [6]:
(6/6)*(5/6)*(4/6)*(3/6)

0.2777777777777778

## 3

In [7]:
(6/6)*(5/6)*(4/6)

0.5555555555555556

## 2

In [8]:
(6/6)*(5/6)

0.8333333333333334

## 1

In [9]:
(6/6)

1.0

***
#### References
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[1] authorNames, referenceTitle, locationWebsiteBookVideo, dateMonthYear.

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[10]

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[100]

&nbsp;&nbsp;&nbsp;[1000]
***
#### END

#### Task 3
&#x1F4CC;
***

***
#### References
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[1] authorNames, referenceTitle, locationWebsiteBookVideo, dateMonthYear.

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[10]

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[100]

&nbsp;&nbsp;&nbsp;[1000]
***
#### END

#### Task 4
&#x1F4CC;
***

***
#### References
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[1] authorNames, referenceTitle, locationWebsiteBookVideo, dateMonthYear.

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[10]

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[100]

&nbsp;&nbsp;&nbsp;[1000]
***
#### END

***
#### Appendix A

```python
# "Classy" solution...
class Bag(dict):
    def __init__(self, alist):
        for elem in alist:
            self.add(elem)

    def add(self, elem):
        self[elem] = self.get(elem, 0) + 1

    def __str__(self):
        out = ['%-8s %3d' % (key, val)
                        for (key, val) in sorted(self.items())]
        return '\n'.join(out)

print Bag('foo bar bar foo glag foo'.split())
```

[...] E. Wiethoff, "Bag Sum In Many Programming Languages," [wiki.c2](https://web.archive.org/web/20201107143701/https://wiki.c2.com/?BagSumInManyProgrammingLanguages), n.d.

#### Appendix B

```python
# Pythonic ListComprehension and loop solution...
bag = 'foo bar bar foo glag foo'.split()
bagsum = dict([(elem, bag.count(elem)) for elem in bag])
for key, val in sorted(bagsum.items()):
    print '%-8s %3d' % (key, val)
```

[...] E. Wiethoff, "Bag Sum In Many Programming Languages," [wiki.c2](https://web.archive.org/web/20201107143701/https://wiki.c2.com/?BagSumInManyProgrammingLanguages), n.d.

#### Appendix C

```python
words = 'foo bar bar foo glag foo'.split()
words.sort()
prev, count = words[0], 1
for word in words[1:]:
    if word == prev:
        count += 1
    else:
        print prev, count
        prev, count = word, 1
print prev, count
```

[...] Anonymous, "Bag Sum In Many Programming Languages," [wiki.c2](https://web.archive.org/web/20201107143701/https://wiki.c2.com/?BagSumInManyProgrammingLanguages), n.d.
***
#### END

***
# END